File: solutions/lua/grains/1/grains.lua
local grains = {}

function grains.square(n)
	return 2 ^ (n - 1)
end

function grains.total()
	local sum = 0
	for sq = 1, 64 do
		sum = sum + grains.square(sq)
	end
	return sum
end

return grains

File: solutions/haskell/grains/1/src/Grains.hs
module Grains (square, total) where

import Data.Maybe (mapMaybe)

square :: Integer -> Maybe Integer
square n
  | n >= 1 && n <= 64 = Just (2 ^ (n -1))
  | otherwise = Nothing

total :: Integer
total = sum . mapMaybe square $ [1 .. 64]
